storm-client/src/jvm/org/apache/storm/streams/processors/CoGroupByKeyProcessor.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.streams.processors;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import org.apache.storm.streams.Pair;
+import org.apache.storm.streams.operations.PairValueJoiner;
+import org.apache.storm.streams.tuple.Tuple3;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * co-group by key implementation
+ */
+public class CoGroupByKeyProcessor<K, V1, V2> extends BaseProcessor<Pair<K, ?>> implements BatchProcessor {
+    private final PairValueJoiner<Collection<V1>, Collection<V2>> valueJoiner;
+    private final String firstStream;
+    private final String secondStream;
+    private final List<Pair<K, V1>> firstRows = new ArrayList<>();
+    private final List<Pair<K, V2>> secondRows = new ArrayList<>();
+
+    public CoGroupByKeyProcessor(String firstStream, String secondStream, PairValueJoiner<Collection<V1>, Collection<V2>> valueJoiner) {
+        this.valueJoiner = valueJoiner;
+        this.firstStream = firstStream;
+        this.secondStream = secondStream;
+    }
+
+    @Override
+    public void execute(Pair<K, ?> input, String sourceStream) {
+        K key = input.getFirst();
+        if (sourceStream.equals(firstStream)) {
+            V1 val = (V1) input.getSecond();
+            Pair<K, V1> pair = Pair.of(key, val);
+            firstRows.add(pair);
+        } else if (sourceStream.equals(secondStream)) {
+            V2 val = (V2) input.getSecond();
+            Pair<K, V2> pair = Pair.of(key, val);
+            secondRows.add(pair);
+        }
+        if (!context.isWindowed()) {
+            joinAndForward(firstRows, secondRows);
+        }
+
+    }
+
+    @Override
+    public void finish() {
+        joinAndForward(firstRows, secondRows);
+        firstRows.clear();
+        secondRows.clear();
+    }
+
+    private void joinAndForward(List<Pair<K, V1>> firstRows, List<Pair<K, V2>> secondRows) {
+        for (Tuple3<K, Collection<V1>, Collection<V2>> res : join(getJoinTable(firstRows), getJoinTable(secondRows))) {
```

Comment:
Once you maintain the two multi-maps, you don't need the join step. You could iterate over them directly and forward the entries.
```java
        mm1.asMap().forEach((key, values) -> {
            context.forward(Pair.of(key, Pair.of(values, mm2.removeAll(key))));
        });

        mm2.asMap().forEach((key, values) -> {
             context.forward(Pair.of(key, Pair.of(mm1.get(key), values)));
        });
```